A notification keeps popping up

Everytime I press on the coding tab a notification appears saying: "A request has failed. See the output for more information.".
![image](https://user-images.githubusercontent.com/67760452/89932232-af1a0a80-dbe3-11ea-825a-5f1230b25213.png)

Not sure if this is of use @bryphe, but here are my error logs, setup and steps:

Setup:
* Windows 10,
* matklad.rust-analyzer extension installed from within Onivim 2
* rust-analyzer binaries downloaded and placed inside one of my PATH folders

Steps:
1. Open Onivim 2 in a rust project folder
1. Open a .rs file in the editor

Log:
```
[ERROR]  +530ms Oni2.Service.FileWatcher : 'oni://Welcome': no such file or directory
[ERROR]  +365ms Oni2.Extension.ClientStore : {"$isError":true,"name":"Error","message":"Assertion failed (active editor 'onivim.editor:0' does not exist)","stack":"Error: Assertion failed (active editor 'onivim.editor:0' does not exist)\n    at Object.ok (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\common\\assert.js:14:19)\n    at ExtHostDocumentsAndEditors.$acceptDocumentsAndEditorsDelta (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\workbench\\api\\common\\extHostDocumentsAndEditors.js:79:28)\n    at RPCProtocol._doInvokeHandler (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\workbench\\services\\extensions\\common\\rpcProtocol.js:336:31)\n    at RPCProtocol._invokeHandler (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\workbench\\services\\extensions\\common\\rpcProtocol.js:321:49)\n    at RPCProtocol._receiveRequest (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\workbench\\services\\extensions\\common\\rpcProtocol.js:248:36)\n    at RPCProtocol._receiveOneMessage (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\workbench\\services\\extensions\\common\\rpcProtocol.js:176:30)\n    at C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\workbench\\services\\extensions\\common\\rpcProtocol.js:66:56\n    at Emitter.fire (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\common\\event.js:457:42)\n    at BufferedEmitter.fire (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\parts\\ipc\\common\\ipc.net.js:362:35)\n    at C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\workbench\\services\\extensions\\node\\extensionHostProcessSetup.js:145:41\n    at Emitter.fire (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\common\\event.js:457:42)\n    at BufferedEmitter.fire (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\parts\\ipc\\common\\ipc.net.js:362:35)\n    at PersistentProtocol._receiveMessage (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\parts\\ipc\\common\\ipc.net.js:581:37)\n    at C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\parts\\ipc\\common\\ipc.net.js:455:83\n    at Emitter.fire (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\common\\event.js:457:42)\n    at ProtocolReader.acceptChunk (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\parts\\ipc\\common\\ipc.net.js:188:37)\n    at C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\parts\\ipc\\common\\ipc.net.js:157:61\n    at Socket.listener (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\parts\\ipc\\node\\ipc.net.js:17:40)\n    at Socket.emit (events.js:315:20)\n    at addChunk (_stream_readable.js:295:12)\n    at readableAddChunk (_stream_readable.js:271:9)\n    at Socket.Readable.push (_stream_readable.js:212:10)\n    at Pipe.onStreamRead (internal/stream_base_commons.js:186:23)"}
[ERROR]    +2ms Oni2.Extension.ClientStore : {"$isError":true,"name":"Error","message":"Assertion failed (document 'file:///oni%3A//Welcome' does not exist)","stack":"Error: Assertion failed (document 'file:///oni%3A//Welcome' does not exist)\n    at Object.ok (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\common\\assert.js:14:19)\n    at ExtHostDocumentsAndEditors.$acceptDocumentsAndEditorsDelta (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\workbench\\api\\common\\extHostDocumentsAndEditors.js:71:32)\n    at RPCProtocol._doInvokeHandler (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\workbench\\services\\extensions\\common\\rpcProtocol.js:336:31)\n    at RPCProtocol._invokeHandler (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\workbench\\services\\extensions\\common\\rpcProtocol.js:321:49)\n    at RPCProtocol._receiveRequest (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\workbench\\services\\extensions\\common\\rpcProtocol.js:248:36)\n    at RPCProtocol._receiveOneMessage (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\workbench\\services\\extensions\\common\\rpcProtocol.js:176:30)\n    at C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\workbench\\services\\extensions\\common\\rpcProtocol.js:66:56\n    at Emitter.fire (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\common\\event.js:457:42)\n    at BufferedEmitter.fire (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\parts\\ipc\\common\\ipc.net.js:362:35)\n    at C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\workbench\\services\\extensions\\node\\extensionHostProcessSetup.js:145:41\n    at Emitter.fire (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\common\\event.js:457:42)\n    at BufferedEmitter.fire (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\parts\\ipc\\common\\ipc.net.js:362:35)\n    at PersistentProtocol._receiveMessage (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\parts\\ipc\\common\\ipc.net.js:581:37)\n    at C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\parts\\ipc\\common\\ipc.net.js:455:83\n    at Emitter.fire (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\common\\event.js:457:42)\n    at ProtocolReader.acceptChunk (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\parts\\ipc\\common\\ipc.net.js:188:37)\n    at C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\parts\\ipc\\common\\ipc.net.js:157:61\n    at Socket.listener (C:\\Program Files\\Onivim2\\node\\node_modules\\@onivim\\vscode-exthost\\out\\vs\\base\\parts\\ipc\\node\\ipc.net.js:17:40)\n    at Socket.emit (events.js:315:20)\n    at addChunk (_stream_readable.js:295:12)\n    at readableAddChunk (_stream_readable.js:271:9)\n    at Socket.Readable.push (_stream_readable.js:212:10)\n    at Pipe.onStreamRead (internal/stream_base_commons.js:186:23)"}
[ERROR]  +366ms Oni2.Extension.ClientStore : Unhandled workspace message - $checkExists: [[{"$mid":1,"external":"file:///<folder with two separate rust projects, fails to find cargo.toml (still get all the other errors even if opening a specific project)>
[ERROR] +4663ms Oni2.Extension.ClientStore : I tried the following decoders but they all failed:
  "string" decoder: Expected a string, but got null 
  "int" decoder: Expected an int, but got null 
  "float" decoder: Expected a float, but got null  
[ERROR]    +1ms Oni2.Extension.ClientStore : Unhandled method: $registerNavigateTypeSupport - Args: [7]
[ERROR]    +0ms Oni2.Extension.ClientStore : Unhandled method: $registerQuickFixSupport - Args: [8,[{"$serialized":true,"language":"rust","scheme":"file"}],{"providedKinds":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite"]},"rust-analyzer"]
[ERROR]    +2ms Oni2.Extension.ClientStore : Unhandled method: $registerFoldingRangeProvider - Args: [15,[{"$serialized":true,"language":"rust","scheme":"file"}]]
[ERROR]    +0ms Oni2.Extension.ClientStore : Unhandled method: $registerSelectionRangeProvider - Args: [16,[{"$serialized":true,"language":"rust","scheme":"file"}]]
[ERROR]    +0ms Oni2.Extension.ClientStore : Unhandled method: $registerCallHierarchyProvider - Args: [17,[{"$serialized":true,"language":"rust","scheme":"file"}]]
[ERROR]    +1ms Oni2.Extension.ClientStore : Unhandled method: $registerDocumentSemanticTokensProvider - Args: [18,[{"$serialized":true,"language":"rust","scheme":"file"}],{"tokenModifiers":["documentation","declaration","definition","static","abstract","deprecated","readonly","constant","controlFlow","injected","mutable","consuming","unsafe","attribute"],"tokenTypes":["comment","keyword","string","number","regexp","operator","namespace","type","struct","class","interface","enum","typeParameter","function","member","property","macro","variable","parameter","label","attribute","boolean","builtinType","enumMember","escapeSequence","formatSpecifier","generic","lifetime","punctuation","selfKeyword","typeAlias","union","unresolvedReference"]},null]
[ERROR]    +1ms Oni2.Extension.ClientStore : Unhandled method: $registerDocumentRangeSemanticTokensProvider - Args: [19,[{"$serialized":true,"language":"rust","scheme":"file"}],{"tokenModifiers":["documentation","declaration","definition","static","abstract","deprecated","readonly","constant","controlFlow","injected","mutable","consuming","unsafe","attribute"],"tokenTypes":["comment","keyword","string","number","regexp","operator","namespace","type","struct","class","interface","enum","typeParameter","function","member","property","macro","variable","parameter","label","attribute","boolean","builtinType","enumMember","escapeSequence","formatSpecifier","generic","lifetime","punctuation","selfKeyword","typeAlias","union","unresolvedReference"]}]
[ERROR]    +1ms Oni2.Extension.ClientStore : Unhandled method: $setLanguageConfiguration - Args: [22,"ra_syntax_tree",{"brackets":[["[",")"]]}]
[ERROR]  +602ms Oni2.Extension.ClientStore : I tried the following decoders but they all failed:
  "string" decoder: Expected a string, but got null 
  "int" decoder: Expected an int, but got null 
  "float" decoder: Expected a float, but got null  
[ERROR]    +9ms Oni2.Extension.ClientStore : I tried the following decoders but they all failed:
  "string" decoder: Expected a string, but got null 
  "int" decoder: Expected an int, but got null 
  "float" decoder: Expected a float, but got null  
[ERROR]   +91ms Oni2.Extension.ClientStore : Expected an int, but got 96.15384615384616
[ERROR] +1562ms Oni2.Extension.ClientStore : Unknown message - type: 6
[ERROR] +12.46s Oni2.Extension.ClientStore : Unable to parse OutputService method: $reveal
```

Looks like there are a couple issues:

This: 
```
[ERROR]  +602ms Oni2.Extension.ClientStore : I tried the following decoders but they all failed:
  "string" decoder: Expected a string, but got null 
  "int" decoder: Expected an int, but got null 
  "float" decoder: Expected a float, but got null  
```

is due to our status bar $setEntry parsing - the `priority` field happens to be null, but we expect an integer.

After that is unblocked, I see this in the LSP logs:
```
WARN [9/2/2020, 2:08:17 PM]: LSP request failed {
  method: 'rust-analyzer/inlayHints',
  param: {
    textDocument: { uri: 'file:///d%3A%5Chello-rust%5Csrc%5Cmain.rs' }
  },
  error: [ResponseError: url is not a file
  	at handleResponse (c:\Users\bryphe\AppData\Local\Oni2\extensions\matklad.rust-analyzer-0.2.297\out\src\main.js:3413:48)
  	at processMessageQueue (c:\Users\bryphe\AppData\Local\Oni2\extensions\matklad.rust-analyzer-0.2.297\out\src\main.js:3238:17)
  	at Immediate.<anonymous> (c:\Users\bryphe\AppData\Local\Oni2\extensions\matklad.rust-analyzer-0.2.297\out\src\main.js:3222:13)
  	at processImmediate (internal/timers.js:456:21)] {
    code: -32603,
    data: undefined
  }
}
```

Looks like there may be a bug with how we are sending the file URL

--- a/src/Core/Uri.re
+++ b/src/Core/Uri.re
@@ -70,30 +70,6 @@ type t = {
   authority: [@default None] option(string),
 };
 
-let encode = uri =>
-  Json.Encode.(
-    obj([
-      ("$mid", Json.Encode.int(1)), // Magic marshaling id for Uri
-      ("scheme", uri.scheme |> Scheme.encode),
-      ("path", uri.path |> string),
-      ("query", uri.query |> nullable(string)),
-      ("authority", uri.authority |> nullable(string)),
-    ])
-  );
-
-let decode = {
-  Json.Decode.(
-    obj(({field, _}) =>
-      {
-        scheme: field.required("scheme", Scheme.decode),
-        path: field.required("path", string),
-        query: field.optional("query", string),
-        authority: field.optional("authority", string),
-      }
-    )
-  );
-};
-
 module Internal = {
   let isDriveLetter = (candidate: Char.t) => {
     let aCode = Char.code('A');
@@ -157,7 +133,16 @@ let fromScheme = (~scheme: Scheme.t, ~authority=?, ~query=?, path: string) => {
 };
 
 let fromMemory = path => fromScheme(~scheme=Scheme.Memory, path);
-let fromPath = path => fromScheme(~scheme=Scheme.File, path);
+let fromPath = path => {
+
+  // On Windows - we need to normalize backward slashes to forward slashes,
+  // to be in sync with the logic from vscode:
+  // https://github.com/onivim/vscode-exthost/blob/c7df89c1cf0087ca5decaf8f6d4c0fd0257a8b7a/src/vs/base/common/uri.ts#L306
+  // Also see: https://github.com/onivim/oni2/issues/2282
+  let backSlashRegex = Str.regexp("\\\\");
+  let normalizePath = p => Str.global_replace(backSlashRegex, "/", p);
+  fromScheme(~scheme=Scheme.File, normalizePath(path));
+};
 
 let toString = ({scheme, authority, path, query}: t) => {
   let schemeStr = Scheme.toString(scheme);
@@ -195,3 +180,27 @@ let toFileSystemPath = (uri: t) => {
 };
 
 let getScheme = (uri: t) => uri.scheme;
+
+let encode = uri =>
+  Json.Encode.(
+    obj([
+      ("$mid", Json.Encode.int(1)), // Magic marshaling id for Uri
+      ("scheme", uri.scheme |> Scheme.encode),
+      ("path", uri.path |> string),
+      ("query", uri.query |> nullable(string)),
+      ("authority", uri.authority |> nullable(string)),
+    ])
+  );
+
+let decode = {
+  Json.Decode.(
+    obj(({field, _}) =>
+      {
+        scheme: field.required("scheme", Scheme.decode),
+        path: field.required("path", string),
+        query: field.optional("query", string),
+        authority: field.optional("authority", string),
+      }
+    )
+  );
+};

--- a/src/Core/Uri.re
+++ b/src/Core/Uri.re
@@ -134,14 +134,13 @@ let fromScheme = (~scheme: Scheme.t, ~authority=?, ~query=?, path: string) => {
 
 let fromMemory = path => fromScheme(~scheme=Scheme.Memory, path);
 let fromPath = path => {
-
   // On Windows - we need to normalize backward slashes to forward slashes,
   // to be in sync with the logic from vscode:
   // https://github.com/onivim/vscode-exthost/blob/c7df89c1cf0087ca5decaf8f6d4c0fd0257a8b7a/src/vs/base/common/uri.ts#L306
   // Also see: https://github.com/onivim/oni2/issues/2282
-  let backSlashRegex = Str.regexp("\\\\");
-  let normalizePath = p => Str.global_replace(backSlashRegex, "/", p);
-  fromScheme(~scheme=Scheme.File, normalizePath(path));
+  let path = Sys.win32 ? Utility.Path.normalizeBackSlashes(path) : path;
+
+  fromScheme(~scheme=Scheme.File, path);
 };
 
 let toString = ({scheme, authority, path, query}: t) => {

--- a/src/Core/Utility/Path.re
+++ b/src/Core/Utility/Path.re
@@ -14,6 +14,11 @@ let hasTrailingSeparator = path => {
   };
 };
 
+let normalizeBackSlashes = {
+  let backSlashRegex = Str.regexp("\\\\");
+  path => Str.global_replace(backSlashRegex, "/", path);
+};
+
 let toRelative = (~base, path) => {
   let base = hasTrailingSeparator(base) ? base : base ++ Filename.dir_sep;
   Str.replace_first(Str.regexp_string(base), "", path);

--- a/src/Core/Utility/ResultEx.re
+++ b/src/Core/Utility/ResultEx.re
@@ -23,3 +23,7 @@ let tap = f =>
       ok;
     }
   | Error(_) as err => err;
+
+let value = (~default) => fun
+| Ok(v) => v
+| Error(_) => default;

--- a/src/Core/Utility/ResultEx.re
+++ b/src/Core/Utility/ResultEx.re
@@ -24,6 +24,7 @@ let tap = f =>
     }
   | Error(_) as err => err;
 
-let value = (~default) => fun
-| Ok(v) => v
-| Error(_) => default;
+let value = (~default) =>
+  fun
+  | Ok(v) => v
+  | Error(_) => default;

--- a/src/Exthost/Msg.re
+++ b/src/Exthost/Msg.re
@@ -1240,7 +1240,7 @@ module StatusBar = {
       let%bind alignmentNumber =
         alignmentJson |> Internal.decode_value(Decode.int);
       let alignment = alignmentNumber |> intToAlignment;
-      let%bind priority = priorityJson |> Internal.decode_value(Decode.int);
+      let%bind priority = priorityJson |> Internal.decode_value(nullable(Decode.int)) |> Result.map(Option.value(~default=0));
       Ok(
         SetEntry({
           id,

--- a/src/Exthost/Msg.re
+++ b/src/Exthost/Msg.re
@@ -1240,7 +1240,10 @@ module StatusBar = {
       let%bind alignmentNumber =
         alignmentJson |> Internal.decode_value(Decode.int);
       let alignment = alignmentNumber |> intToAlignment;
-      let%bind priority = priorityJson |> Internal.decode_value(nullable(Decode.int)) |> Result.map(Option.value(~default=0));
+      let%bind priority =
+        priorityJson
+        |> Internal.decode_value(nullable(Decode.int))
+        |> Result.map(Option.value(~default=0));
       Ok(
         SetEntry({
           id,

--- a/test/Core/UriTests.re
+++ b/test/Core/UriTests.re
@@ -28,7 +28,13 @@ describe("Uri", ({describe, _}) => {
     test("adds slash for windows-style path", ({expect, _}) => {
       let uri = Uri.fromPath("C:/test");
       expect.string(Uri.toString(uri)).toEqual("file:///c:/test");
-    })
+    });
+    if (Sys.win32) {
+      test("#2282 - Windows: backward slashes should get normalized to forward slashes", ({expect, _}) => {
+        let uri = Uri.fromPath("C:\\hello-rust\\src\\main.rs");
+      expect.string(Uri.toString(uri)).toEqual("file:///c:/hello-rust/src/main.rs");
+      });
+    }
   });
   describe("toFileSystemPath", ({test, _}) => {
     test("round-trip windows-style path", ({expect, _}) => {

--- a/test/Core/UriTests.re
+++ b/test/Core/UriTests.re
@@ -30,11 +30,15 @@ describe("Uri", ({describe, _}) => {
       expect.string(Uri.toString(uri)).toEqual("file:///c:/test");
     });
     if (Sys.win32) {
-      test("#2282 - Windows: backward slashes should get normalized to forward slashes", ({expect, _}) => {
+      test(
+        "#2282 - Windows: backward slashes should get normalized to forward slashes",
+        ({expect, _}) => {
         let uri = Uri.fromPath("C:\\hello-rust\\src\\main.rs");
-      expect.string(Uri.toString(uri)).toEqual("file:///c:/hello-rust/src/main.rs");
+        expect.string(Uri.toString(uri)).toEqual(
+          "file:///c:/hello-rust/src/main.rs",
+        );
       });
-    }
+    };
   });
   describe("toFileSystemPath", ({test, _}) => {
     test("round-trip windows-style path", ({expect, _}) => {

--- a/test/Exthost/Test.re
+++ b/test/Exthost/Test.re
@@ -3,6 +3,7 @@ open Exthost_Extension;
 open Exthost;
 open Exthost_TestLib;
 open Oni_Core;
+open Oni_Core.Utility;
 
 module Log = (val Timber.Log.withNamespace("Test"));
 
@@ -92,7 +93,7 @@ let startWithExtensions =
       ~onError=errorHandler,
       (),
     )
-    |> ResultEx.tap_error(msg => prerr_endline(msg))
+    |> ResultEx.tapError(msg => prerr_endline(msg))
     |> Result.get_ok;
 
   let processHasExited = ref(false);

--- a/test/Exthost/lib/Node.re
+++ b/test/Exthost/lib/Node.re
@@ -1,4 +1,5 @@
 open Oni_Core;
+open Oni_Core.Utility;
 module Log = (val Timber.Log.withNamespace("ExtHost.TestLib.Node"));
 
 let spawn = (~env=[], ~onExit, args) => {
@@ -27,6 +28,6 @@ let spawn = (~env=[], ~onExit, args) => {
     nodeFullPath,
     [nodeFullPath, ...args],
   )
-  |> ResultEx.tap_error(msg => prerr_endline(Luv.Error.strerror(msg)))
+  |> ResultEx.tapError(msg => prerr_endline(Luv.Error.strerror(msg)))
   |> Result.get_ok;
 };

--- a/test/Exthost/lib/ResultEx.re
+++ b/test/Exthost/lib/ResultEx.re
@@ -1,7 +1,0 @@
-let tap_error = f =>
-  fun
-  | Ok(_) as v => v
-  | Error(msg) as v => {
-      f(msg);
-      v;
-    };

